submariner/multiple-active-gateways.md

```diff
@@ -0,0 +1,100 @@
+# Multiple Active Gateways
+
+## Summary
+
+NOTE: The design details are still a work in progress, so this is currently a draft proposal.
+
+Submariner currently only allows a single gateway to be active at any one time in a given cluster.
+This enhancement proposes that there be an option to allow more than one gateway to be active in
+a given cluster to enhance both performance and resiliency. This feature is sometimes referred to as
+“active-active gateways”; however, the goal is to allow two or more actove gateways to be used.
+
+## Proposal
+
+The proposal is to optionally allow multiple gateways to be active at the same time in a given cluster.
+
+When only a single active gateway is used, all inter-cluster traffic must flow through that one gateway,
+which creates a performance bottleneck. Enabling more than one node to be used as a gateway can improve
+the performance linearly in the number of gateways used.
+
+Additionally, when the active gateway fails in the current implementation, work must be done to switch
+over to the new gateway including the establishment of new tunnels (both intra-cluster and inter-cluster)
+and new routes, which takes time. With the proposed design, It is expected that the fail-over times can
+be significantly improved if additional gateways are operational. Furthermore, it is expected that
+traffic flowing through operational gateways will continue to flow while the traffic from the failed
+gateway is redirected to the operational gateways. Both properties reduce the impact of a failure.
+
+## Design Details
+
+### Single Gateway
+
+In the current single gateway Submariner implementation, intra-cluster VXLAN tunnels are established by
+the non-gateway nodes to the gateway node, and each gateway establishes an inter-cluster tunnel to each
+other gateway in the clusterset (with the tunnel type determined by the cable driver in use) as shown
+below.
+
+![SingleGateway](./images/single-gateway.png)
+
+### Multiple Active Gateway Design
+
+The proposal for supporting multiple active gateway nodes is shown in the following diagram and described below.
+
+![MultiGateway](./images/multi-gateway.png)
+
+* Each non-gateway node will establish a VXLAN tunnel to each gateway node.
```

Comment:
Do we have any understand of the performance impact with the configuration of so many more tunnels per clusterset?

Reply:
This is a good point and something that should be covered.  We don't have hard numbers.  However, the goal and expectation are that this will improve forwarding performance. Initial setup could take a little longer for each new endpoint (peer cluster), but it shouldn't be too big of a hit.  However, this could increase the total number of tunnels for a clusterset significantly.  For N nodes, we currently have N(N-1) tunnels. If there are M gateways on each node, we multiply that by 2M^2.  So, I'm thinking that for large deployments, we'd only want to use multiple gateways in a set of high speed "core" clusters, and then connect "leaf" clusters in with fewer connections to a subset of core clusters.  However, this will require that the gateways have "transit node" functionality and that we use dynamic routing, so that's for a future enhancement :-).

Reply:
So, in summary, I do have some scalability concerns for large deployments if we don't also make other enhancements. We'll have to analyze the impact.